pkg/sources/forward/data_forward.go

```diff
@@ -518,10 +520,12 @@ func (isdf *DataForward) applyTransformer(ctx context.Context, readMessage *isb.
 			}
 			continue
 		} else {
-			// if we do not get a time from Transformer, we set it to the time from (N-1)th vertex
 			for index, m := range writeMessages {
 				m.ID = fmt.Sprintf("%s-%s-%d", readMessage.ReadOffset.String(), isdf.vertexName, index)
-				if m.EventTime.IsZero() {
+				// if the message is dropped by the source transformer, we assign the original event time to the message.
+				// the original event time is the event time of the message before the source transformer.
+				// this way, even when the message is dropped, the event time is still preserved and be used for watermark calculation.
+				if slices.Contains(m.Tags, sourcetransformer.DROP) {
```

Comment:
I think we might have to check for IsZero condition, too, or else the watermark might unintendedly progress far ahead.

Reply:
Thanks, Vigith. Do you mean that zero event time might be unintentionally assigned to the transformed messages? if that's the case, do you mean far behind(in the past) or far ahead(in the future) ?

Reply:
also, will checking for `slices.Contains` in the hot code bath be detrimental to performance?